dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -1312,61 +1312,85 @@ private List<Event> handleLKMMIntrinsic(FunctionCall call) {
+                final Expression value = expressions.makeCast(p2, bytes);
+                result.add(EventFactory.Linux.newLKMMStore(p0, value, mo.equals(Tag.Linux.MO_MB) ? Tag.Linux.MO_ONCE : mo));
                 if (mo.equals(Tag.Linux.MO_MB)) {
                     result.add(EventFactory.Linux.newMemoryBarrier());
                 }
             }
-            case "__LKMM_XCHG" -> {
-                checkArgument(p2 instanceof IntLiteral, "No support for variable memory order.");
-                mo = Tag.Linux.intToMo(((IntLiteral) p2).getValueAsInt());
-                result.add(EventFactory.Linux.newRMWExchange(p0, reg, p1, mo));
+            case "__LKMM_xchg" -> {
+                checkArguments(4, call);
+                final Type bytes = toLKMMAccessSize(p1);
+                final String mo = toLKMMMemoryOrder(p3);
+                final Register dummy = call.getFunction().newRegister(bytes);
+                final Expression value = expressions.makeCast(p2, bytes);
+                result.add(EventFactory.Linux.newRMWExchange(p0, dummy, value, mo));
+                result.add(EventFactory.newLocal(reg, expressions.makeCast(dummy, reg.getType())));
             }
-            case "__LKMM_CMPXCHG" -> {
-                checkArgument(p3 instanceof IntLiteral, "No support for variable memory order.");
-                mo = Tag.Linux.intToMo(((IntLiteral) p3).getValueAsInt());
-                result.add(EventFactory.Linux.newRMWCompareExchange(p0, reg, p1, p2, mo));
+            case "__LKMM_cmpxchg" -> {
+                checkArguments(6, call);
+                final Type bytes = toLKMMAccessSize(p1);
+                final String mo = toLKMMMemoryOrder(p4);
+                final Register dummy = call.getFunction().newRegister(bytes);
+                final Expression value = expressions.makeCast(p2, bytes);
+                final Expression expectation = expressions.makeCast(p3, bytes);
+                result.add(EventFactory.Linux.newRMWCompareExchange(p0, dummy, expectation, value, mo));
+                result.add(EventFactory.newLocal(reg, expressions.makeCast(dummy, reg.getType())));
             }
-            case "__LKMM_ATOMIC_FETCH_OP" -> {
-                checkArgument(p2 instanceof IntLiteral, "No support for variable memory order.");
-                mo = Tag.Linux.intToMo(((IntLiteral) p2).getValueAsInt());
-                checkArgument(p3 instanceof IntLiteral, "No support for variable operator.");
-                op = IntBinaryOp.intToOp(((IntLiteral) p3).getValueAsInt());
-                result.add(EventFactory.Linux.newRMWFetchOp(p0, reg, p1, op, mo));
+            case "__LKMM_atomic_fetch_op" -> {
+                checkArguments(5, call);
+                final Type bytes = toLKMMAccessSize(p1);
+                final String mo = toLKMMMemoryOrder(p3);
+                final IntBinaryOp op = toLKMMOperation(p4);
+                final Register dummy = call.getFunction().newRegister(bytes);
```

Comment:
I think you can use `newUniqueRegister` with a base name like "__lkmm_temp/dummy".
Generally, we should reduce the usages of the generic dummy name, ideally getting rid of it altogether (eventually).